--- chromium-79.0.3945.36.orig/content/common/user_agent.cc
+++ chromium-79.0.3945.36/content/common/user_agent.cc
@@ -141,8 +141,7 @@ base::StringPiece GetFrozenUserAgent(boo
 
 std::string BuildUserAgentFromProduct(const std::string& product) {
   std::string os_info;
-  base::StringAppendF(&os_info, "%s%s", GetUserAgentPlatform().c_str(),
-                      BuildOSCpuInfo(false).c_str());
+  base::StringAppendF(&os_info, "%sLinux x86_64", GetUserAgentPlatform().c_str());
   return BuildUserAgentFromOSAndProduct(os_info, product);
 }
 
