--- chromium-80.0.3987.87.orig/third_party/angle/src/compiler/translator/InfoSink.h
+++ chromium-80.0.3987.87/third_party/angle/src/compiler/translator/InfoSink.h
@@ -92,7 +92,16 @@ class TInfoSinkBase
             stream.precision(8);
             stream << f;
         }
-        sink.append(stream.str());
+
+        // Hack to work around a bug where negative floating point values
+        // are rendered like '.0.5' instead of '-0.5'
+        std::string res(stream.str());
+
+        if (signbit(f)) { // test if f is negative
+            res[0] = '-';
+        }
+
+        sink.append(res);
         return *this;
     }
     // Write boolean values as their names instead of integral value.
--- chromium-80.0.3987.87.orig/third_party/skia/src/sksl/SkSLString.cpp
+++ chromium-80.0.3987.87/third_party/skia/src/sksl/SkSLString.cpp
@@ -240,7 +240,12 @@ String to_string(double value) {
     if (needsDotZero) {
         buffer << ".0";
     }
-    return String(buffer.str().c_str());
+
+    std::string ret(buffer.str());
+    if (signbit(value) && ret[0] == '.') {
+        ret[0] = '-';
+    }
+    return String(ret.c_str());
 }

 SKSL_INT stoi(const String& s) {
