--- chromium-79.0.3945.79.orig/chrome/common/chrome_paths_linux.cc
+++ chromium-79.0.3945.79/chrome/common/chrome_paths_linux.cc
@@ -91,7 +91,7 @@ bool GetDefaultUserDataDirectory(base::F
 #if BUILDFLAG(GOOGLE_CHROME_BRANDING)
   *result = config_dir.Append("google-chrome" + GetChannelSuffixForDataDir());
 #else
-  *result = config_dir.Append("chromium");
+  *result = config_dir.Append("ungoogled-chromium");
 #endif
   return true;
 }
