--- llvm/lib/clang/10.0.0/include/altivec.h.orig
+++ llvm/lib/clang/10.0.0/include/altivec.h
@@ -2909,7 +2909,21 @@ static __inline__ vector double __ATTRS_o_ai vec_cpsgn(vector double __a,

 /* vec_cts */

-#ifdef __VSX__
+#if defined(__VSX__) && defined(__cplusplus)
+#define vec_cts(__a, __b)                                                      \
+  _Generic((__a), vector float                                                 \
+           : __builtin_altivec_vctsxs((vector float)(__a), (__b)),             \
+	     __v4sf                                                 \
+           : __builtin_altivec_vctsxs((vector float)(__a), (__b)),             \
+             vector double                                                     \
+           : __extension__({                                                   \
+             vector double __ret =                                             \
+                 (vector double)(__a) *                                        \
+                 (vector double)(vector unsigned long long)((0x3ffULL + (__b)) \
+                                                            << 52);            \
+             __builtin_convertvector(__ret, vector signed long long);          \
+           }))
+#elif defined(__VSX__)
 #define vec_cts(__a, __b)                                                      \
   _Generic((__a), vector float                                                 \
            : __builtin_altivec_vctsxs((vector float)(__a), (__b)),             \
